File: C#/Dijkstra.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;

class Program
{
    const int INF = int.MaxValue;

    // Priority Queue using SortedSet (compatible with older C#)
    class Node : IComparable<Node>
    {
        public int Vertex;
        public int Distance;

        public Node(int v, int d)
        {
            Vertex = v;
            Distance = d;
        }

        public int CompareTo(Node other)
        {
            if (Distance != other.Distance)
                return Distance.CompareTo(other.Distance);
            return Vertex.CompareTo(other.Vertex);
        }
    }

    static void Dijkstra(List<Tuple<int, int>>[] graph, int source)
    {
        int n = graph.Length;
        int[] dist = new int[n];
        for (int i = 0; i < n; i++)
            dist[i] = INF;

        SortedSet<Node> pq = new SortedSet<Node>();
        dist[source] = 0;
        pq.Add(new Node(source, 0));

        while (pq.Count > 0)
        {
            Node node = pq.Min;
            pq.Remove(node);

            int u = node.Vertex;

            foreach (var edge in graph[u])
            {
                int v = edge.Item1;
                int weight = edge.Item2;

                if (dist[u] != INF && dist[u] + weight < dist[v])
                {
                    // Remove old entry if exists
                    pq.Remove(new Node(v, dist[v]));

                    dist[v] = dist[u] + weight;
                    pq.Add(new Node(v, dist[v]));
                }
            }
        }

        Console.WriteLine("\nShortest distances from source:");
        for (int i = 0; i < n; i++)
        {
            if (dist[i] == INF)
                Console.WriteLine("Node " + i + ": unreachable");
            else
                Console.WriteLine("Node " + i + ": " + dist[i]);
        }
    }

    static void Main()
    {
        Console.Write("Enter number of vertices and edges: ");
        string[] firstLine = Console.ReadLine().Split();
        int n = int.Parse(firstLine[0]);
        int m = int.Parse(firstLine[1]);

        List<Tuple<int, int>>[] graph = new List<Tuple<int, int>>[n];
        for (int i = 0; i < n; i++)
            graph[i] = new List<Tuple<int, int>>();

        Console.WriteLine("Enter edges (u v weight) [0-based indexing]:");
        for (int i = 0; i < m; i++)
        {
            string[] parts = Console.ReadLine().Split();
            int u = int.Parse(parts[0]);
            int v = int.Parse(parts[1]);
            int w = int.Parse(parts[2]);

            graph[u].Add(Tuple.Create(v, w));
            // For undirected graph, uncomment:
            // graph[v].Add(Tuple.Create(u, w));
        }

        Console.Write("Enter source vertex: ");
        int source = int.Parse(Console.ReadLine());

        Stopwatch sw = Stopwatch.StartNew();
        Dijkstra(graph, source);
        sw.Stop();

        Console.WriteLine("\nExecution time: " + sw.Elapsed.TotalSeconds + " seconds");
    }
}

File: Rust/Dijkstra.rs
use std::cmp::Reverse;
use std::collections::BinaryHeap;
use std::io::{self, Read};
use std::time::Instant;

const INF: i64 = i64::MAX / 4;

fn dijkstra(adj: &Vec<Vec<(usize, i64)>>, src: usize) -> Vec<i64> {
    let n = adj.len();
    let mut dist = vec![INF; n];
    let mut heap = BinaryHeap::new();

    dist[src] = 0;
    heap.push(Reverse((0_i64, src)));

    while let Some(Reverse((d, u))) = heap.pop() {
        if d != dist[u] {
            continue; // stale entry
        }
        for &(v, w) in &adj[u] {
            let nd = d + w;
            if nd < dist[v] {
                dist[v] = nd;
                heap.push(Reverse((nd, v)));
            }
        }
    }

    dist
}

fn main() {
    // Read all input
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let mut it = input.split_whitespace();

    println!("Enter number of vertices and edges:");
    let n: usize = match it.next() {
        Some(v) => v.parse().unwrap(),
        None => return,
    };
    let m: usize = it.next().unwrap().parse().unwrap();

    let mut adj: Vec<Vec<(usize, i64)>> = vec![Vec::new(); n];

    println!("Enter edges (u v weight) [0-based indexing]:");
    for _ in 0..m {
        let u: usize = it.next().unwrap().parse().unwrap();
        let v: usize = it.next().unwrap().parse().unwrap();
        let w: i64 = it.next().unwrap().parse().unwrap();
        adj[u].push((v, w));
        // For undirected graph, also add:
        // adj[v].push((u, w));
    }

    println!("Enter source vertex:");
    let src: usize = it.next().unwrap().parse().unwrap();

    let start = Instant::now();
    let dist = dijkstra(&adj, src);
    let duration = start.elapsed();

    println!("\nShortest distances from source {}:", src);
    for i in 0..n {
        if dist[i] == INF {
            println!("Node {}: unreachable", i);
        } else {
            println!("Node {}: {}", i, dist[i]);
        }
    }

    println!(
        "\nExecution time: {:.9} seconds",
        duration.as_secs_f64()
    );
}
